### PR TITLE
Permit noreply outcomes from API Endpoint callback modules

### LINKED ISSUE #4645: Elixir: expand `Ockam.Services.API.Endpoint` to handle multiple API requests concurrently
### Current behavior

Currently `Ockam.Services.API.Endpoint.dispatch/5` always waits for handler to complete before returning `{:reply ...`.

This means that only one request can run on the same API worker at a time. While functions defined in handler may take some time.

### Desired behavior

If handler functions don't have to block each other (as it usually is in API workers), they can instead run concurrently.

Base API worker `Ockam.Services.API.handle_request/2` supports returning `{:noreply, ...}` and replying later. This can be used in `dispatch/5` or `handle_request/2` callback implementations in `Ockam.Services.API.Endpoint`

### PR DESCRIPTION
**Pull request by shanesveller**
## Current behavior

Elixir modules that `use Ockam.Services.API.Endpoint` currently must return an ok-tuple or error-tuple from their request handler functions in order to satisfy the `case` statement of `dispatch/5` [here](https://github.com/build-trust/ockam/blob/47a03863942682b24a1d47e8f4e7862c21a7ffde/implementations/elixir/ockam/ockam_services/lib/services/api/endpoint.ex#L120).

## Proposed changes

In order to allow such downstream modules to perform asynchronous replies out-of-band, it must be allowed to return `:noreply` or `{:noreply, new_state}`. Implementation details to conduct those replies in practice are not included here, though they would likely be built atop `Ockam.Services.API.reply` and `reply_error`.

## Checks

- [X] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [X] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [X] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [X] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository. The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/edit/main/CONTRIBUTORS.csv) file in the github web UI and create a PR, this will mark the commit as verified.

### CODE REVIEW

File: implementations/elixir/ockam/ockam_services/lib/services/api/endpoint.ex [hunk 1]
```diff
@@ -125,6 +125,12 @@ defmodule Ockam.Services.API.Endpoint do
           {:ok, body, new_endpoint_state} ->
```
Comment (hairyhum) [inline implementations/elixir/ockam/ockam_services/lib/services/api/endpoint.ex:125]: Should we change those to `{:reply, :ok, body}` and `{:reply, :ok, body, state}` for consistency?
@polvorin any particular reason why it was changed to `{:ok, body}`? That prevents handlers from returning other status codes btw.

**Review by hairyhum — COMMENTED** [review]
I think we can make it more consistent and more flexible by allowing different status codes by changing the `{:ok, body}` result as well

File: implementations/elixir/ockam/ockam_services/lib/services/api/endpoint.ex (hunk 1 above)
Comment (polvorin) [reply implementations/elixir/ockam/ockam_services/lib/services/api/endpoint.ex:125]: right, as is on the success case the handlers returns the  response body  so no way to indicate a success code other than 200.    So if we want to do things like 201 "Created"  we will need to change this.
Other quirk is that the handler accept a Request{},   but don't return a Response{}  (returns a response body,   or on a upper level,   a status code and response body).
It feels that  a Request{}  -> Response{}  here would make more sense.
 
Regarding the async response.. in experience so far it looks like most things that are implemented behind an endpoint like this *don't*  mutate any internal state in any way and are fine to be run in parallel.   Would it make sense to keep the handlers unaware of all this,  and just be a configuration on the routes?   (run the handler 
 in an `one-at-a-time way` like now,   run the handler in a `freshly spawned fun`,  run the handler in a spawned fun but up to N at a given time at max,  etc).   So the handler remains just as a function from req -> resp.     (of course this won't apply for handlers that _do_ change worker state)

Comment (shanesveller) [reply implementations/elixir/ockam/ockam_services/lib/services/api/endpoint.ex:125]: @hairyhum @polvorin I haven't had enough time to develop the context I would need to speak to or contribute a larger reimagining of this code's design. Would it be okay if we treated those as two separate conversations and discuss the merits or concerns of this change as-is, and make separate time to discuss your points above?

Comment (hairyhum) [reply implementations/elixir/ockam/ockam_services/lib/services/api/endpoint.ex:125]: @polvorin @shanesveller Let's make a follow-up issue with the proposal to support "async handlers" and maybe handler pools as this grows the design scope of this issue.

**Review by hairyhum — DISMISSED** [review]
@shanesveller since we want to review that in the follow-up, it's fine with me if we don't switch to `{:reply, :ok` now.

**Review by mszpakowski — DISMISSED** [review]

**Comment by mszpakowski** [comment]
bats tests are currently failing, and that's the reason I'm waiting with merging

**Review by mszpakowski — APPROVED** [review]

**Pull request merged**